utils/zerovec/src/map/map.rs

```diff
@@ -63,8 +63,13 @@ where
     K: ZeroMapKV<'a> + ?Sized,
     V: ZeroMapKV<'a> + ?Sized,
 {
-    pub(crate) keys: K::Container,
-    pub(crate) values: V::Container,
+    // These need to be public because we need to const-construct `ZeroMap`s for
+    // Crabbake. It's not possible to define a const constructor with the correct
+    // generic bounds without the `const_fn_trait_bound` feature.
+    #[doc(hidden)]
+    pub keys: K::Container,
+    #[doc(hidden)]
+    pub values: V::Container,
```

Comment:
Issue, here and in ZeroMap2d: ZeroMap has really strong invariants on these fields. I'm really nervous adding public fields, even hidden ones, because it can result in code with no unsafe blocks that could have unsafe behavior.

@Manishearth Is there a way to avoid this?

Reply:
Yeah, I'm against making these fields public, even if it's doc(hidden).

I do not want crabbake to cause semantic changes to our other types; if that is happening we should be rethinking things.

There is _one_ potential hack which is that we define `BakeableZeroMap` with the same layout and public fields, and transmute.

I'll see if I can get a carve-out for associated types in the const trait restrictions.

Reply:
I asked Oli about this and he said it's on nightly already, so we should be getting this in a couple releases. Perhaps we can temporarily do the unsafe thing? Or temporarily do the pub/hidden thing, but I'd prefer to not tweak the main API

Reply:
Basically on nightly we already can have trait bounds in const fns, we just can't call trait methods, which we don't need anyway

Reply:
Shane has suggested the transmute as well, but the problem there is that type information will not flow through. If we have a field `map: ZeroMap<'data, str, str>`, then the expression `core::mem::transmute(BakeableZeroMap(VarZeroVec::from_bytes(&[...]), VarZeroVec::from_bytes(&[...]))` will not work for it, because the `str` type information doesn't flow through the transmute. Crabbake relies heavily on inferred types, and this would require all values to bake their types and to explicitly add generic types everywhere.

Would it be possible to remove the type bounds from the `ZeroMap` struct? This would allow us to add an `unsafe const` constructor. However I think because the fields are associated types of the generics this won't work.

Reply:
Fixed by #1926